rs/nervous_system/root/src/change_canister.rs

```
@@ -256,24 +304,51 @@ async fn install_code(request: ChangeCanisterRequest) -> ic_cdk::api::call::Call
     let canister_id = canister_id.get();
     let sender_canister_version = Some(ic_cdk::api::canister_version());
 
-    let install_code_args = InstallCodeArgs {
-        mode,
-        canister_id,
-        wasm_module,
-        arg,
-        compute_allocation,
-        memory_allocation,
-        sender_canister_version,
-    };
-    // Warning: despite dfn_core::call returning a Result, it actually traps when
-    // the callee traps! Use the public cdk instead, which does not have this
-    // issue.
-    ic_cdk::api::call::call(
-        Principal::try_from(IC_00.get().as_slice()).unwrap(),
-        "install_code",
-        (&install_code_args,),
-    )
-    .await
+    if let Some(ChunkedCanisterWasm {
```

There will need to be validation somewhere that chunked_canister_wasm and wasm_module are mutually exclusive, I think.

---

I was thinking about it, too, but couldn't convince myself that this was necessary. 

Currently, if there are some bytes in `wasm_module` and `chunked_canister_wasm`, then the former would be used and the latter ignored. 

Are you sure it's better to reject requests that specify more data than we need? 

Either way, I'm happy to add a unit test to make sure this behavior is well specified.

---

I think the principle of least surprise should apply.  One would not expect to be able to submit two mutually exclusive options into an 
API and not get an error.  If the API silently ignores one, this could be confusing.  

It might be a good change, for future forward compatibility, to make the new field actually support both as an enum, and then translate the old field into the enum as part of the validation, so that we can eventually retire the old field and have the type structure enforce the either-or nature of the arguments.

---

> so that we can eventually retire the old field and have the type structure enforce the either-or nature of the arguments

I see your point. Although, to me it's not obvious that the old field should be retired in this case. The old field could still be used for upgrading smaller canisters using a potentially simpler end-user flow that doesn't require a separate operation to upload the WASM chunks.

---

The old field could be retired b/c it could be superseded by an enum field that does the same thing.  

So you could have "wasm_to_install" be 
```
enum WasmToInstall {
    Chunked(ChunkedCanisterWasm)
    Wasm(bytes)
}
```
or something like that, so that the types could enforce the exclusivity.  You'd still allow for smaller canisters to not chunk their wasms - it would just be that the arguments would make more sense.

---

Thanks for clarifying, I agree that having a single enum field 

```
    wasm: WasmToInstall,
```

would be a bit cleaner design than 

```
    wasm_module: Vec<u8>,
    chunked_canister_wasm: ChunkedCanisterWasm,
```

After staring into the code again, I'm not sure what you mean by "The old field could be retired" in this case, as in terms of Candid, it is a non-optional `blob` field. 

So we would have two options: 
1. Break Root's compatibility (sounds risky, even the set of allowed clients for Root is known, i.e., the Governances)
2. Keep a forever validation code to check that the blob is not populated. 

If we go with the 2nd approach, then IMHO the the enum doesn't fully serve its purpose. 

Since this is an internal API, maybe in this case we could just follow your original suggestion and validate the exclusiveness of two fields?